File: python/SynC.py
import numpy as np
import scipy


def ComputeNeighborhood(p, eps, D):
    N = D.shape[0]

    neighborhood = []

    x = D[p]
    for j in range(N):
        y = D[j]
        dist = np.linalg.norm(x - y)
        if dist < eps:
            neighborhood.append(j)

    return np.array(neighborhood)


def ComputeKNN(p, k, D):
    N = D.shape[0]

    distances = np.full(k, np.inf)
    neighborhood = np.zeros(k)

    x = D[p]
    for j in range(N):
        if j == p:
            continue
        y = D[j]
        dist = np.linalg.norm(x - y)
        if dist < distances[0]:
            l = 1
            while l < k and dist < distances[l]:
                distances[l - 1] = distances[l]
                neighborhood[l - 1] = neighborhood[l]
                l += 1

            distances[l - 1] = dist
            neighborhood[l - 1] = j

    return neighborhood


def UpdatePoint(x, N_p, D_current, D_next):
    if len(N_p) > 0:
        D_next[x] = D_current[x] + 1 / len(N_p) * np.sum([
            np.sin(D_current[y] - D_current[x])  # todo why sin?
            # D_current[y] - D_current[x]
            for y in N_p
        ], axis=0)
    else:
        D_next[x] = D_current[x]


def ComputeLocationOrder(x, N_x, D):
    N = D.shape[0]

    if len(N_x) == 0:
        return 1

    r_c = 1 / len(N_x) * np.sum([
        np.exp(-np.abs(np.linalg.norm(D[x] - D[y])))
        for y in N_x
    ])

    return r_c


def synCluster(D):
    N = D.shape[0]

    C = np.full(N, -1)
    Cs = []

    cl = 0

    for i in range(N):
        if C[i] == -1:
            in_cluster = False
            C_cl = []
            for j in range(N):
                if i != j and np.allclose(D[i], D[j]):
                    C[j] = cl
                    C_cl.append(j)
                    in_cluster = True
            if in_cluster:
                C[i] = cl
                C_cl.append(i)
                cl += 1
                Cs.append(C_cl)

    return Cs


def Outliers(D, Cl):
    N = D.shape[0]

    O = []
    for i in range(N):
        for C_j in Cl:
            if i not in C_j:
                O.append(i)

    return np.array(O)


def NN(k, D):
    N = D.shape[0]
    d = D.shape[1]

    distances = np.full(k, np.inf)

    avg = 0

    for i in range(N):
        x = D[i]
        distances[:] = np.inf
        for j in range(N):
            if j == i:
                continue
            y = D[j]
            dist = np.linalg.norm(x - y)
            if dist < distances[0]:
                l = 1
                while l < k and dist < distances[l]:
                    distances[l - 1] = distances[l]
                    l += 1

                distances[l - 1] = dist
        avg += distances[0]#np.mean(distances)

    return avg / N


def K(x):
    return (2 * np.pi) ** (-1 / 2) * np.exp(-x ** 2 / 2)


def f_hat(x, D, h):
    N = D.shape[0]
    d = D.shape[1]

    sum = 0
    for y in range(N):
        prod = 1
        for i in range(d):
            prod *= 1 / h[i] * K((D[x, i] - D[y, i]) / h[i])
        sum += prod
    return 1 / N * sum


def f_hat_in_C(x, D, h, C):
    N = D.shape[0]
    d = D.shape[1]

    sum = 0
    for y in C:
        prod = 1
        for i in range(d):
            prod *= 1 / h[i] * K((D[x, i] - D[y, i]) / h[i])
        sum += prod
    return 1 / len(C) * sum


# def pdf(x, D, h, sum_f_hat):
#     return f_hat(x, D, h) / sum_f_hat


def L(D, M):  # todo something seems to be wrong with the cost function???
    N = D.shape[0]
    d = D.shape[1]
    C = M[0]
    O = M[1]
    K = len(C)
    p_i = d

    L_M = 0
    for i in range(K):
        L_M += len(C[i]) * np.log2(N / len(C[i]))
        L_M += p_i / 2 * np.log2(len(C[i]))

    # for x in range(len(O)):
    #     L_M += 1 * np.log2(N)
    #     L_M += p_i / 2 * np.log2(1)

    sigma = np.array([np.var(D[:, j]) for j in range(d)])
    IQR = np.array([scipy.stats.iqr(D[:, j]) for j in range(d)])
    h = np.array([0.9 * (N ** (-1 / (d + 4))) * min(sigma[j], IQR[j] / 1.34) for j in range(d)])
    sum_f_hat = np.sum([f_hat(y, D, h) for y in range(N)])

    L_D_given_M = 0
    for i in range(K):
        # sigma = np.array([np.var(D[C[i], j]) for j in range(d)])
        # IQR = np.array([scipy.stats.iqr(D[C[i], j]) for j in range(d)])
        # h = np.array([0.9 * (N ** (-1 / (d + 4))) * min(sigma[j], IQR[j] / 1.34) for j in range(d)])
        # sum_f_hat = np.sum([f_hat_in_C(y, D, h, C[i]) for y in C[i]])
        # sum_f_hat = np.sum([f_hat(y, D[C[i],:], h, ) for y in range(len(C[i]))])
        for x in C[i]:
            pdf_x = f_hat(x, D, h) / sum_f_hat
            # pdf_x = f_hat_in_C(x, D, h, C[i]) / sum_f_hat

            L_D_given_M += np.log2(pdf_x)

    # for x in O:
    #     pdf_x = f_hat(x, D, h) / sum_f_hat
    #     L_D_given_M += np.log2(pdf_x)

    L_D_given_M *= -1
    print("L_M:", L_M)
    print("L_D_given_M:", L_D_given_M)
    return L_M + L_D_given_M


def DynamicalClustering(D, eps, lam=1 - 1e-3, call=None):
    D_current = D.copy()
    D_next = D.copy()
    call(D_next)
    loopFlag = True
    while loopFlag:
        r_local = 0
        for p in range(len(D)):
            N_p = ComputeNeighborhood(p, eps, D_current)
            UpdatePoint(p, N_p, D_current, D_next)
            r_p = ComputeLocationOrder(p, N_p, D_current)
            r_local += r_p
            # print("r_p:", r_p)

        r_local /= len(D)

        print("r_local:", r_local)
        if call is not None:
            call(D_next)

        if r_local > lam:  # todo this why of terminating is strange!
            loopFlag = False
            Cl = synCluster(D_next)
            Ol = Outliers(D_next, Cl)
            Ml = (Cl, Ol)
        tmp = D_next
        D_next = D_current
        D_current = tmp

    return Ml


def Sync(D, k, call=None):
    l = 0
    eps = NN(k, D)
    delta_eps = NN(k + 1, D) - eps
    global_synchronization = False
    M = []
    coding_cost = []
    while not global_synchronization:
        print("l:", l, "eps", eps)
        M.append(DynamicalClustering(D, eps, call=call))
        if len(M[l][0]) == 1:
            global_synchronization = True
        print("computing cost:")
        coding_cost.append(L(D, M[l]))
        print(coding_cost[-1])
        eps += delta_eps
        l += 1

    l_star = np.argmin(coding_cost)  # forall l
    M_star = M[l_star]
    return M_star

File: src/algorithms/SynC.cpp
//
// Created by mrjak on 14-07-2021.
//

#include <cmath>
#include <ctime>
#include <iostream>
#include <ATen/ATen.h>
#include "omp.h"

#include "SynC.h"
#include "../utils/CPU_math.h"
#include "../utils/Timer.h"
#include "../structure/RTree.h"

neighborhood ComputeNeighborhood(int p, float eps, float *D, int n, int d) {
    neighborhood N;

    float *x = &D[p * d];
    for (int j = 0; j < n; j++) {
        float *y = &D[j * d];
        float dist = distance(x, y, d);
        if (dist <= eps)
            N.push_back(j);
    }

    return N;
}

void UpdatePoint(int x, const neighborhood &N_p, float *D_current, float *D_next, int n, int d) {
    for (int l = 0; l < d; l++) {
        D_next[x * d + l] = D_current[x * d + l];

        float sum = 0.;
        for (int i: N_p) {
            sum += sin(D_current[i * d + l] - D_current[x * d + l]);
        }
        D_next[x * d + l] += sum / N_p.size();
    }
}

float ComputeLocationOrder(int x, const neighborhood &N_x, float *D, int n, int d) {
    if (N_x.size() == 0)
        return 1.;

    float sum = 0.;
    for (int i: N_x) {
        sum += exp(-abs(distance(&D[x * d], &D[i * d], d)));
    }
    //    if (x == 0) {
    //        printf("CPU: r_c=%f/%d=%f\n", sum, N_x.size(), sum / N_x.size());
    //    }
    float r_c = sum / N_x.size();

    return r_c;
}

clustering synCluster(float *D, int n, int d, float eps) {
    int *C = full(n, -1);
    clustering Cs;

    int cl = 0;

    for (int i = 0; i < n; i++) {
        if (C[i] == -1) {
            bool in_cluster = false;
            cluster C_cl;
            C_cl.push_back(i);
            for (int j = 0; j < n; j++) { // we can do this better
                if (i != j && distance(&D[i * d], &D[j * d], d) <= eps) { //todo dist not needed!
                    C[j] = cl;
                    C_cl.push_back(j);
                    in_cluster = true;
                }
            }
            if (in_cluster) {
                C[i] = cl;
                Cs.push_back(C_cl);
                cl += 1;
            }
        }
    }

    delete C;
    return Cs;
}

at::Tensor synCluster_v2(float *D, int n, int d, float eps) {
//    int *C = full(n, -1);

    std::vector<int> C(n);
    for (int i = 0; i < n; i++) {
        C[i] = -1;
    }

//    clustering Cs;

    int cl = 0;

    for (int i = 0; i < n; i++) {
        if (C[i] == -1) {
            bool in_cluster = false;
//            cluster C_cl;
//            C_cl.push_back(i);
            for (int j = 0; j < n; j++) { // we can do this better
                if (i != j && distance(&D[i * d], &D[j * d], d) <= eps) { //todo dist not needed!
                    C[j] = cl;
//                    C_cl.push_back(j);
                    in_cluster = true;
                }
            }
            if (in_cluster) {
                C[i] = cl;
//                Cs.push_back(C_cl);
                cl += 1;
            }
        }
    }

//    delete C;
//    return Cs;

    at::Tensor C_ = at::zeros(n, at::kInt);
    for (int i = 0; i < n; i++) {
        C_[i] = C[i];
    }

    return C_;
}

outliers Outliers(float *D, int n, int d, clustering &C) {
    outliers O;
    for (int i = 0; i < n; i++) {
        for (cluster C_j: C) {
            if (!contains(C_j, i)) {
                O.push_back(i);
            }
        }
    }
    return O;
}

float NN(float *D, int n, int d, int k) {

    float *dists = full(k, INF);

    float avg = 0;

    for (int i = 0; i < n; i++) {
        float *x = &D[i * d];
        set_all(dists, k, INF);
        for (int j = 0; j < n; j++) {
            if (i == j)
                continue;
            float *y = &D[j * d];
            float dist = distance(x, y, d);
            if (dist < dists[0]) {
                int l = 1;
                while (l < k && dist < dists[l]) {
                    dists[l - 1] = dists[l];
                    l += 1;
                }

                dists[l - 1] = dist;
            }
        }
        avg += dists[0]; //mean(dists, k);
    }
    return avg / n;
}

float K(float x) {
    return pow(2 * PI, -1. / 2.) * exp(-pow(x, 2) / 2.);
}

float f_hat(int x, float *D, int n, int d, float *h) {
    float sum = 0.;
    for (int y = 0; y < n; y++) {
        float prod = 1.;
        for (int i = 0; i < d; i++) {
            prod *= 1. / h[i] * K((D[x * d + i] - D[y * d + i]) / h[i]);
        }
        sum += prod;
    }
    return 1. / n * sum;
}

float L(float *D, const int n, const int d, clustering &M) {

    clustering C = M;

    float L_M = 0.;
    for (int i = 0; i < C.size(); i++) {
        L_M += C[i].size() * log2(((float) n) / ((float) C[i].size()));
        L_M += d / 2. * log2((float) C[i].size());
    }

    //    printf("Got to: line %d in file %s\n", __LINE__, __FILE__);

    const float *sigma = variance(D, n, d);
    const float *IQR = iqr(D, n, d);
    float *h = new float[d];
    for (int j = 0; j < d; j++) {
        h[j] = 0.9 * pow(n, (-1 / (d + 4))) * min(sigma[j], IQR[j] / 1.34);
    }
    float sum_f_hat = 0;
    for (int y = 0; y < d; y++) {
        sum_f_hat += f_hat(y, D, n, d, h);
    }

    float L_D_given_M = 0;
    for (int i = 0; i < C.size(); i++) {
        for (auto x: C[i]) {
            float pdf_x = f_hat(x, D, n, d, h) / sum_f_hat;
            L_D_given_M += std::log2f(pdf_x);
        }
    }
    L_D_given_M *= -1;

    delete sigma;
    delete IQR;
    delete h;

    return L_M + L_D_given_M;
}

float L_Clust(float *D, const int n, const int d, clustering &M) {

    clustering C = M;

    float L_M = 0.;
    for (int i = 0; i < C.size(); i++) {
        L_M += C[i].size() * log2(((float) n) / ((float) C[i].size()));
        L_M += d / 2. * log2((float) C[i].size());
    }

    //    printf("Got to: line %d in file %s\n", __LINE__, __FILE__);

    const float *sigma = variance(D, n, d);
    const float *IQR = iqr(D, n, d);
    float *h = new float[d];
    for (int j = 0; j < d; j++) {
        h[j] = 0.9 * pow(n, (-1 / (d + 4))) * min(sigma[j], IQR[j] / 1.34);
    }

    float L_D_given_M = 0;
    for (int i = 0; i < C.size(); i++) {

        float sum_f_hat = 0;
        for (auto y: C[i]) {
            sum_f_hat += f_hat(y, D, n, d, h);
        }

        for (auto x: C[i]) {
            float pdf_x = f_hat(x, D, n, d, h) / sum_f_hat;
            L_D_given_M += std::log2f(pdf_x);
        }
    }
    L_D_given_M *= -1;

    delete sigma;
    delete IQR;
    delete h;

    return L_M + L_D_given_M;
}

std::vector <at::Tensor> DynamicalClustering(float *D, int n, int d, float eps, float lam) {
    Timer<6> timer;

    timer.start_stage_time(0);
    float *D_current = clone(D, n * d);
    float *D_next = clone(D, n * d);

    float r_local = 0.;
    timer.end_stage_time(0);
    while (r_local < lam) {
        timer.start_itr_time();
        timer.start_stage_time(2);
        r_local = 0.;
        for (int p = 0; p < n; p++) {
            const neighborhood N_t = ComputeNeighborhood(p, eps, D_current, n, d);
            UpdatePoint(p, N_t, D_current, D_next, n, d);
//            const neighborhood N_tp = ComputeNeighborhood(p, eps, D_current, n, d);
            float r_p = ComputeLocationOrder(p, N_t, D_current, n, d);
            r_local += r_p;
        }

//        for (int p = 0; p < n; p++) {//this is more correct, but not actually what they are doing
//            const neighborhood N_tp = ComputeNeighborhood(p, eps, D_next, n, d);
//            float r_p = ComputeLocationOrder(p, N_tp, D_next, n, d);
//            r_local += r_p;
//        }

        r_local /= n;
        printf("r_local: %f\n", r_local);

        float *tmp = D_next;
        D_next = D_current;
        D_current = tmp;
        timer.end_stage_time(2);
        timer.end_itr_time();
    }

    timer.start_stage_time(4);
    at::Tensor C = synCluster_v2(D_current, n, d, eps);
    timer.end_stage_time(4);

    std::vector <at::Tensor> result;
    result.push_back(C);

    at::Tensor itr_times = timer.get_itr_times();
    result.push_back(itr_times);

    at::Tensor stage_times = timer.get_stage_times();
    result.push_back(stage_times);

    at::Tensor space = at::zeros(1, at::kInt);
    space[0] = 0;
    result.push_back(space);

    delete D_current;
    delete D_next;

    return result;
}

std::vector <at::Tensor> DynamicalClustering_parallel(float *D, int n, int d, float eps, float lam) {


    Timer<6> timer;

//    omp_set_num_threads(12);
    printf("Max number of threads: %d\n", omp_get_max_threads());

    timer.start_stage_time(0);
    float *D_current = clone(D, n * d);
    float *D_next = clone(D, n * d);
    float * rs = new float[omp_get_max_threads()];

    float r_local = 0.;
    timer.end_stage_time(0);
    while (r_local < lam) {
        timer.start_itr_time();
        timer.start_stage_time(2);
        r_local = 0.;

        for (int i =0;i<omp_get_max_threads();i++){
            rs[i] = 0.;
        }

#pragma omp parallel for
        for (int p = 0; p < n; p++) {
//            printf("Number of threads: %d/%d\n", omp_get_thread_num(),omp_get_num_threads());
            const neighborhood N_t = ComputeNeighborhood(p, eps, D_current, n, d);
            UpdatePoint(p, N_t, D_current, D_next, n, d);
//            const neighborhood N_tp = ComputeNeighborhood(p, eps, D_current, n, d);
            float r_p = ComputeLocationOrder(p, N_t, D_current, n, d);
//            r_local += r_p;
            rs[omp_get_thread_num()] += r_p;
        }
        r_local = 0.;
        for (int i =0;i<omp_get_max_threads();i++){
            r_local +=rs[i];
        }


        r_local /= n;
        printf("r_local: %f\n", r_local);

        float *tmp = D_next;
        D_next = D_current;
        D_current = tmp;
        timer.end_stage_time(2);
        timer.end_itr_time();
    }

    timer.start_stage_time(4);
    at::Tensor C = synCluster_v2(D_current, n, d, eps);
    timer.end_stage_time(4);

    std::vector <at::Tensor> result;
    result.push_back(C);

    at::Tensor itr_times = timer.get_itr_times();
    result.push_back(itr_times);

    at::Tensor stage_times = timer.get_stage_times();
    result.push_back(stage_times);

    at::Tensor space = at::zeros(1, at::kInt);
    space[0] = 0;
    result.push_back(space);

    delete D_current;
    delete D_next;

    return result;
}

//clustering SynC(float *h_D, int n, int d, int k) {
//    int l = 0;
//    float eps = NN(h_D, n, d, k);
//    float delta_eps = NN(h_D, n, d, k + 1) - eps;
//    bool global_synchronization = false;
//    std::vector <clustering> M;
//    std::vector<float> coding_cost;
//    while (!global_synchronization) {
//        clustering C = DynamicalClustering(h_D, n, d, eps, 1 - 1e-3);
//        M.push_back(C);
//        if (M[l].size() == 1) {
//            global_synchronization = true;
//        }
//        float cost = L(h_D, n, d, M[l]);
//        coding_cost.push_back(cost);
//        eps += delta_eps;
//        l += 1;
//    }
//
//    int l_star = arg_min(coding_cost); // forall l
//    clustering M_star = M[l_star];
//    return M_star;
//}


std::vector <at::Tensor> FDynamicalClustering(float *D, int n, int d, float eps, float lam, int B) {
    Timer<6> timer;

    timer.start_stage_time(0);
    float *D_current = clone(D, n * d);
    float *D_next = clone(D, n * d);
    timer.end_stage_time(0);

    RTree tree(d, B);
    for (int p = 0; p < n; p++) {
        tree.insert(new std::pair<int, float *>(p, &D_current[p * d]));
    }

    float r_local = 0.;
    while (r_local < lam) {
        timer.start_itr_time();
        timer.start_stage_time(2);
        r_local = 0.;

        for (int p = 0; p < n; p++) {
            const neighborhood N_t = tree.range(eps, &D_current[p * d]);//ComputeNeighborhood(p, eps, D_current, n, d);
            UpdatePoint(p, N_t, D_current, D_next, n, d);
            float r_p = ComputeLocationOrder(p, N_t, D_current, n, d);
            r_local += r_p;
        }
        timer.end_stage_time(2);

        timer.start_stage_time(1);
        for (int p = 0; p < n; p++) {
            tree.remove(new std::pair<int, float *>(p, &D_current[p * d]));
            tree.insert(new std::pair<int, float *>(p, &D_next[p * d]));
        }
        timer.end_stage_time(1);

        r_local /= n;
        printf("r_local: %f\n", r_local);

        float *tmp = D_next;
        D_next = D_current;
        D_current = tmp;
        timer.end_itr_time();
    }

    timer.start_stage_time(4);
    at::Tensor C = synCluster_v2(D_current, n, d, eps);
    timer.end_stage_time(4);

    std::vector <at::Tensor> result;
    result.push_back(C);

    at::Tensor itr_times = timer.get_itr_times();
    result.push_back(itr_times);

    at::Tensor stage_times = timer.get_stage_times();
    result.push_back(stage_times);

    at::Tensor space = at::zeros(1, at::kInt);
    space[0] = 0;
    result.push_back(space);

    return result;
}